py/scheduler.c

```diff
@@ -60,22 +60,26 @@ static inline bool mp_sched_empty(void) {
 void mp_handle_pending(bool raise_exc) {
     if (MP_STATE_VM(sched_state) == MP_SCHED_PENDING) {
         mp_uint_t atomic_state = MICROPY_BEGIN_ATOMIC_SECTION();
-        mp_obj_t obj = MP_STATE_VM(mp_pending_exception);
```

Comment:
I'm pretty sure the existing code is safe if there is no threading (eg on bare-metal without threading).  Because in that case IRQs are the only async thing and they cannot change the state from pending to not pending (but IRQs could schedule more things).

But it'll be tricky/messy to make this fix conditional on `MICROPY_PY_THREAD`.  And I understand why you've made it check twice for `MP_SCHED_PENDING`, because the `MICROPY_BEGIN_ATOMIC_SECTION` call may be expensive.

So, I guess it's good to have this fix as-is.